### DISCUSSION TITLE
Pybricks Balancer Example

### DISCUSSION THREAD

**Discussion by GAuzinger** [pybricks/support#231] [Q&A]
Hello,
Im new to all of this and tried to run the script on my TechnicHub. I changed every InventorHub to TechnicHub (thats how far my programming experience is), what else do i have to change?
Big Shoutouts to Pybricks - looks promising!!!

**Comment by laurensvalk** [comment]
Yes, that's all. Depending on which [sensor](https://docs.pybricks.com/en/latest/pupdevices/index.html) you have, you can pick either `ColorDistanceSensor` or `ColorSensor`.

Let us know how it goes :)

Did you build it the same as in the video?

**Reply by GAuzinger** [reply]
Hello laurensvalk, yes it is working, but i´ve learned much till it worked (how to "start" the balancer for example)! I built a kind-of-version of the model in the video... now im looking forward to try another sensors and commands for another colors. It is very funny and very exciting too, really!

**Reply by laurensvalk** [reply]
Cool!

**Reply by Werner-G** [reply]
Hallo, I'm struggeling with the TechnicHub. Can you please share your code with us? Sounds difficult to find the correct ratings for the duty cycle.

**Reply by GAuzinger** [reply]
i think that was it - its been a long time ago. Most important is that the robot is "in balance" when starting, that means it stands "upright" and that there is no "movement".... I also built the model a little bit different, i am looking for a photo and will post it for you.... 

```python
from pybricks.hubs import TechnicHub
from pybricks.pupdevices import Motor, ColorSensor
from pybricks.parameters import Port, Direction
from [pybricks.tools](http://pybricks.tools/) import wait, StopWatch
from pybricks.geometry import Axis

# Initialize motors.
left = Motor(Port.C, Direction.COUNTERCLOCKWISE)
right = Motor(Port.D)
left.reset_angle(0)
right.reset_angle(0)

# Initialize hub and color sensor.
# You can also use the TechnicHub and/or the ColorDistanceSensor instead.
hub = TechnicHub()
sensor = ColorSensor(Port.B)

# Initialize position buffer for speed calculation.
DT = 5
window = int(300/DT)
buf = [0] * window
idx = 0
angle = 0
DRIVE_SPEED = 300
PAUSE = 5000

# Timer to generate reference position.
watch = StopWatch()

while True:
    # Get angular rate and estimated angle.
    rate = hub.imu.angular_velocity(Axis.Y)
    angle += rate*DT/1000

    # Get motor position.
    position = (left.angle() + right.angle()) / 2

    # Calculate motor speed.
    speed = (position - buf[idx]) / (window*DT) * 1000
    buf[idx] = position
    idx = (idx + 1) % window

    # Calculate reference position, which just grows linearly with time.
    reference = -max(watch.time() - PAUSE, 0) / 1000 * DRIVE_SPEED

    # Calculate duty cycle.
    duty = 0.018*rate + 19*angle + 0.45*(position - reference) + 0.16*speed

    # Account for battery level and type.
    duty *= 7200/hub.battery.voltage()

    # Calculate steering.
    reflection = sensor.reflection()
    steering = (reflection - 28)*0.6

    # Apply duty cycle for balancing and steering.
    left.dc(duty + steering)
    right.dc(duty - steering)

    # Wait some time.
    wait(DT)
```

**Reply by GAuzinger** [reply]
https://photos.app.goo.gl/BaQp5bUtxLVZArnj6

**Reply by Werner-G** [reply]
Thank you Gernot, you are right: it is very difficult to start the balancer. After just one little moment, it runs with full speed and never stops. But I'm using another motor than you - the 88013 Powered Up L-motor. Even when I reduce the speed in line 39, I have the same effect. Any idea how to solve the problem?

**Reply by GAuzinger** [reply]
No at first glance i have no answer. How did you install the motor? Are they mounted vertical or horizontal? In my humble opinion: "the hub has to be as high as possible above the ground", so try to mount the motors vertical using - for example a "BEAM 3M Ø4.85 W. FORK" Elementnummer: 6278412 Designnummer: 49137 to get the hub as high as possible.... I´ll try it as soon as possible with this motors, but this will take some time (holidays)...... Cheers & Keep on bricking!

**Reply by Werner-G** [reply]
https://photos.app.goo.gl/m9AKFBNiRpWvMP426
https://photos.app.goo.gl/uywoQe4j3Zpg6CqH7

Thanks for your help & suggestions. As you can see, I already use the fork beam. With bigger tires, my balancer is now running, but I still think the speed is much too high. Is it necessary modify some of the parameters in the code?

**Reply by GAuzinger** [reply]
Hello Werner,
Sure you have to change the parameters... just try & error! I made a 3-stud-high (in my case)  platform that stabilized the robot before starting. So robot stands still and upright - program is started - platform shall be quickly removed!!
So in your case i would suggest: Get the sensor "straight" - not "skew", place the black platform (if your line is black) that the robot stands upright, start the program, remove the platform, watch it perform... and then change parameters and try again. Your construction looks good - the simpler, the better. 
Cheers & Keep on bricking!
Gernot

**Reply by GAuzinger** [reply]
And consider: your wheels are much more smaller then the ones i used - so the parameters **must** be changed...
